source/base/ThreadPool.ooc

```diff
@@ -1,111 +1,203 @@
+				this _resultTask _mutex unlock()
+				break
+			}
+			this _resultTask _mutex unlock()
+			Time sleepMilli(100)
+		}
+		status
+	}
+	wait: func ~default (defaultValue: T) -> T {
+		status := this wait()
+		status ? this _resultTask _result[T] : defaultValue
+	}
+	cancel: override func -> Bool {
+		status := false
+		this _resultTask _mutex lock()
+		if (this _resultTask _state == _TaskState Unfinished) {
+			this _resultTask _state = _TaskState Cancelled
+			status = true
+		}
+		this _resultTask _mutex unlock()
+		status
+	}
+}
+
 ThreadPool: class {
+	_globalmutex := Mutex new()
 	_threads: Thread[]
-	_jobs := BlockedQueue<ThreadJob> new()
-	_mutex := Mutex new()
+	_threadMutexes: Mutex[]
+	_tasks := BlockedQueue<_Task> new()
 	_allFinishedCondition := WaitCondition new()
 	_activeJobs: Int
 	_threadCount: Int
 	threadCount ::= this _threadCount
-
 	init: func (threadCount := 4) {
 		this _threadCount = threadCount
 		this _threads = Thread[threadCount] new()
+		this _threadMutexes = Mutex[threadCount] new()
 		for (i in 0 .. threadCount) {
-			this _threads[i] = Thread new(|| threadLoop())
+			this _threadMutexes[i] = Mutex new()
+			this _threads[i] = Thread new(|| threadLoop(i))
 			this _threads[i] start()
 		}
 	}
 	free: override func {
-		for (i in 0 .. this _threadCount)
+		for (i in 0 .. this _threadCount) {
 			this _threads[i] free()
+			this _threadMutexes[i] destroy()
+		}
 		gc_free(this _threads data)
-		this _jobs free()
+		this _tasks free()
 		this _allFinishedCondition free()
-		this _mutex destroy()
+		this _globalmutex destroy()
 		super()
 	}
-	threadLoop: func {
+	threadLoop: func (threadId: Int) {
 		while (true) {
-			job := this _jobs wait()
-			job run()
-			this _mutex lock()
+			job := this _tasks wait()
+			job run(this _threadMutexes[threadId])
+			this _globalmutex lock()
 			this _activeJobs -= 1
-			if (this _activeJobs == 0)
+			if (this _activeJobs == 0) {
 				this _allFinishedCondition broadcast()
-			this _mutex unlock()
+			}
+			this _globalmutex unlock()
 		}
 	}
-	_add: func (job: ThreadJob) {
-		this _mutex lock()
+	_add: func (task: _Task) -> Void {
+		this _globalmutex lock()
 		this _activeJobs += 1
-		this _mutex unlock()
-		this _jobs enqueue(job)
+		this _globalmutex unlock()
+		this _tasks enqueue(task)
+	}
+	add: func (action: Func) {
+		task := _Task new(action, this _globalmutex)
+		this _add(task)
+	}
+	addWait: func (action: Func) -> Promise {
+		task := _Task new(action, this _globalmutex)
+		this _add(task)
+		_TaskPromise new(task)
 	}
-	addSynchronized: func (body: Func) -> SynchronizedThreadJob {
-		job := SynchronizedThreadJob new(body)
-		this _add(job)
-		job
+	addWaitResult: func ~result <T> (action: Func -> T) -> ResultPromise<T> {
+		task := _ResultTask<T> new(action, this _globalmutex)
+		this _add(task)
+		_ResultTaskPromise new(task)
 	}
-	add: func (body: Func) { this _add(ThreadJob new(body)) }
-	wait: func (body: Func) { this addSynchronized(body) wait() . free() }
```

Comment:
Why did you remove this? It's nice to have when you want to add something to the pool and synchronize immediately after, without having to deal with waiting on and freeing promises.